src/include/duckdb/storage/compression/alp/algorithm/alp.hpp

```diff
@@ -0,0 +1,397 @@
+
+			//! We try all combinations in search for the one which minimize the compression size
+			for (int8_t exp_idx = AlpTypedConstants<T>::MAX_EXPONENT; exp_idx >= 0; exp_idx--) {
+				for (int8_t factor_idx = exp_idx; factor_idx >= 0; factor_idx--) {
+					idx_t exceptions_count = 0;
+					idx_t non_exceptions_count = 0;
+					uint32_t estimated_bits_per_value = 0;
+					uint64_t estimated_compression_size = 0;
+					int64_t max_encoded_value = NumericLimits<int64_t>::Minimum();
+					int64_t min_encoded_value = NumericLimits<int64_t>::Maximum();
+
+					for (idx_t sample_idx = 0; sample_idx < n_samples; sample_idx++) {
+						T value = sampled_vector[sample_idx];
+
+						T decoded_value;
+						T tmp_encoded_value;
+						int64_t encoded_value;
+
+						tmp_encoded_value =
+						    value * AlpTypedConstants<T>::EXP_ARR[exp_idx] * AlpTypedConstants<T>::FRAC_ARR[factor_idx];
+						encoded_value = NumberToInt64(tmp_encoded_value);
+
+						//! The cast to T is needed to prevent a signed integer overflow
+						decoded_value = static_cast<T>(encoded_value) * AlpConstants::FACT_ARR[factor_idx] *
+						                AlpTypedConstants<T>::FRAC_ARR[exp_idx];
+						if (decoded_value == value) {
+							non_exceptions_count++;
+							max_encoded_value = MaxValue(encoded_value, max_encoded_value);
+							min_encoded_value = MinValue(encoded_value, min_encoded_value);
+							continue;
+						}
+						exceptions_count++;
+					}
+
+					if (non_exceptions_count < 2) { // We skip combinations which yields to almost all exceptions
+						continue;
+					}
+					// Evaluate factor/exponent compression size (we optimize for FOR)
+
+					uint64_t delta =
+					    (static_cast<uint64_t>(max_encoded_value) - static_cast<uint64_t>(min_encoded_value));
+					estimated_bits_per_value = std::ceil(std::log2(delta + 1));
+					estimated_compression_size += n_samples * estimated_bits_per_value;
+					estimated_compression_size +=
+					    exceptions_count * (EXACT_TYPE_BITSIZE + (AlpConstants::EXCEPTION_POSITION_SIZE * 8));
+
+					if ((estimated_compression_size < best_total_bits) ||
+					    // We prefer bigger exponents
+					    (estimated_compression_size == best_total_bits && (best_exponent < exp_idx)) ||
+					    // We prefer bigger factors
+					    ((estimated_compression_size == best_total_bits && best_exponent == exp_idx) &&
+					     (best_factor < factor_idx))) {
+						best_total_bits = estimated_compression_size;
+						best_exponent = exp_idx;
+						best_factor = factor_idx;
+					}
+				}
+			}
+
+			pair<int8_t, int8_t> best_combination = make_pair(best_exponent, best_factor);
+			if (best_k_combinations_hash.count(best_combination)) {
+				best_k_combinations_hash[best_combination] += 1;
+			} else {
+				best_k_combinations_hash[best_combination] = 1;
+			}
+		}
+
+		// Convert our hash pairs to a Combination vector to be able to sort
+		vector<AlpCombination> best_k_combinations;
```

Comment:
A MaxHeap would indeed be a micro-optimisation. However it could be over-engineering? Since this vector is of a small size always (~9 Combinations), and this is done only once per segment. 
It would improve code readability though.
I have added it as a FIXME in a comment

Reply:
Ah then disregard the comment, it wasn't clear to me that the amount of combinations was that trivial 👍

Reply:
Makes sense. I'll remove the FIXME comment